test/host_tests/sal_tcpserver.py

```diff
@@ -0,0 +1,152 @@
+# Copyright 2015 ARM Limited
+#
+# Licensed under the Apache License, Version 2.0
+# See LICENSE file for details.
+
+import sys
+import socket
+import select
+import time
+import threading
+from SocketServer import BaseRequestHandler, TCPServer, _eintr_retry
+from mbed_host_tests import BaseHostTest
+
+
+class TCPServerV4(TCPServer):
+    address_family = socket.AF_INET
+    allow_reuse_address = True
+
+    def __init__(self, server_address, RequestHandlerClass):
+        TCPServer.__init__(self, server_address, RequestHandlerClass)
+        self._shutdown_request = False
+        
+        # watchdog guards against the failure mode when the remote target fails 
+        # to send any packets. If the watchdog reaches the high water mark, the 
+        # server is terminated so as not to leave the server thread unterminated
+        self.watchdog = 0.0
+
+    def serve_forever(self, poll_interval=0.5):
+        """provide an override that can be shutdown from a request handler.
+        The threading code in the BaseSocketServer class prevented this from working
+        even for a non-threaded blocking server.
+        """
+        
+        try:
+            while not self._shutdown_request:
+                r, w, e = _eintr_retry(select.select, [self], [], [], poll_interval)
+                if self in r:
+                    # reset watchdog
+                    self.watchdog = 0.0
+                    self._handle_request_noblock()
+                
+                else:
+                    self.watchdog += poll_interval
+                
+                
+                if self.watchdog > 20.0:
+                    self._shutdown_request = True
+                
+        finally:
+            self._shutdown_request = False
+        
+        
+
+class TCPHandler(BaseRequestHandler):
+    MAX_RX_SIZE = 8192
+    def handle(self):
+        """ One handle per connection
+        """
+        print("HOST: Connection received...\n")
+        while True:
+            try:
+                data = self.request.recv(self.MAX_RX_SIZE)
+            except Exception as e:
+                    print("HOST: detected unexpected exception: %s" % str(e))
+                    break
+                
+            if not data: 
+                break
+            
+            print("HOST: Received %s bytes of data\n" % len(data))
+            
+            # try not shutting down the connection and wait for the remote end to close it
+            if data == "shutdown":
+                print("HOST: Requesting server shutdown\n")
+                # give the remote end time to close its connection, which is part of the test
+                time.sleep(5)
```

Comment:
- this is just to give the remote end time to close the connect without the local host generating an exception when it can read data from the socket, as noted in the comment
- there is little benefit from making it configurable
- it should be a defined value with explanation as to purpose. I will make these changes.